upload/upload.go

```diff
@@ -172,28 +186,71 @@ func getDirPathForThumbnail(recipeID int64) string {
 	return filepath.Join(getDirPathForRecipe(recipeID), "thumbs")
 }
 
-func toResampleFilter(q ImageQualityLevel) imaging.ResampleFilter {
-	switch q {
-	case ImageQualityHigh:
-		return imaging.Box
+func fit(src image.Rectangle, size int) image.Rectangle {
+	// Compute the two possible scale factors.
+	scaleW := float64(size) / float64(src.Dx())
+	scaleH := float64(size) / float64(src.Dy())
+
+	// Pick the *smaller* factor so the whole image stays visible.
+	scale := math.Min(scaleW, scaleH)
+
+	newW := int(math.Round(float64(src.Dx()) * scale))
+	newH := int(math.Round(float64(src.Dy()) * scale))
+	return image.Rect(0, 0, newW, newH)
+}
+
+func cover(src image.Rectangle, size int) image.Rectangle {
+	// Compute the two possible scale factors.
+	scaleW := float64(size) / float64(src.Dx())
+	scaleH := float64(size) / float64(src.Dy())
+
+	// Pick the *larger* factor so the image fills the box.
+	scale := math.Max(scaleW, scaleH)
+
+	newW := int(math.Round(float64(src.Dx()) * scale))
+	newH := int(math.Round(float64(src.Dy()) * scale))
+
+	// Offsets for a centred crop.
+	offsetX := (newW - size) / 2
+	offsetY := (newH - size) / 2
+
+	return image.Rect(
+		offsetX,
+		offsetY,
+		newW + offsetX,
+		newH + offsetY)
+}
+
+func resizeImage(src image.Image, dstW, dstH int, scaler draw.Scaler) *image.RGBA {
+	dst := image.NewRGBA(image.Rect(0, 0, dstW, dstH))
+	scaler.Scale(dst, dst.Bounds(), src, src.Bounds(), draw.Over, nil)
```

Comment:
We don't care about overlaying

```suggestion
	scaler.Scale(dst, dst.Bounds(), src, src.Bounds(), draw.Src, nil)
```